File: app/fantabot_app/api/v1/endpoints/lineup.py
"""Lineup preview and submit — the best fieldable formation for a lega.

**`GET /lineup/plan` never submits. `POST /lineup/submit` is a dry run unless the request
asks to arm, and even then only if `FANTABOT_AUTO_ACT` is set.** Two locks, both named
separately when shut, and `arm` has **no default**: a request that does not say does not act.
That is not paranoia about a typo — it is the property that the operator who armed it is the
one watching, and a browser can be reloaded, restored by a session manager, or left open
overnight.

Both routes call `application/lineup_submit.py`, which is where the eight decisions on this
path live. `GET` used to hand-write the seven reads that build a plan, which is how the app
came to read the format from a different place than the command did.

Mirrors interface/lineup.py's plan path: my_team -> teamLineup_read -> lineup_settings ->
inputs_from_lineup -> plan_lineups, then returns the top PlannedLineup. This is the only
read that hits the live platform (apileague, bearer token), so it degrades open to a
reason when there is no key, no token, or no network — and it never calls
teamLineup_submit.
"""

from __future__ import annotations

from contextlib import contextmanager
from datetime import datetime
from pathlib import Path
from typing import Any

from fantabot.adapters.http.apileague import teamLineup_read
from fantabot.application.arming import ARM, AUTO_ACT
from fastapi import APIRouter
from pydantic import BaseModel

router = APIRouter()


@contextmanager
def _open_store() -> Any:
    """A `TokenStore` on an open session, and the one door out of `GET /lineup/current`.

    A seam rather than four lazy imports inside the route: it is what the route's own tests
    replace, and a name imported inside a function body is one a `monkeypatch.setattr` on
    this module cannot reach. No bearer enters this frame — `TokenStore` resolves it inside
    the adapter, which is `test_app_never_handles_a_plaintext_token`'s whole subject.
    """
    from fantabot.adapters.persistence import database_manager
    from fantabot.adapters.tokens.store import TokenStore
    from fantabot.config import settings
    from fantabot.domain.tokens.crypto import TokenCipher

    with database_manager.get_session() as session:
        yield TokenStore(session, TokenCipher(settings.fantabot_encryption_key))


def _now() -> datetime:
    """The app's clock seam for the lineup path — one, so a harness has one target.

    `interface/lineup.py::_now` is the CLI's, and `tests/domain/asta/test_asta_clock.py`
    counts seams per surface. The value feeds `is_past_deadline`, which compares naive.
    """
    return datetime.now()  # noqa: DTZ005 — naive, as the platform's `mstr` is


class LineupPlayer(BaseModel):
    player_id: int
    nome: str


class LineupPlan(BaseModel):
    found: bool
    #: One of `api/outcomes.LINEUP_PLAN_OUTCOMES`. This route already carried a `reason`,
    #: which made it look solved — but every failure produced the *same* one, "Not
    #: connected, or no lineup available yet", so a network blip and a roster the platform
    #: refuses read identically and only one of them is worth waiting out.
    outcome: str = "planned"
    reason: str | None = None
    module: str = ""
    matchday: int | None = None
    #: Which competition this plan is for, resolved by `build_plans` rather than asked for.
    #: Carried because `GET /lineup/current` needs one and the page has no other way to
    #: learn it — and because "the plan" and "what is saved" are only comparable when both
    #: name the same competition. `build_plans` already returned it and it was discarded.
    competition: int | None = None
    starters: list[LineupPlayer] = []
    bench: list[LineupPlayer] = []


def build_lineup_plan(
    planned: Any, names: dict[int, str], competition: int | None = None
) -> LineupPlan:
    """Map a PlannedLineup + id->name dict to the response (pure)."""
    return LineupPlan(
        found=True,
        outcome="planned",
        module=planned.module,
        matchday=planned.mday,
        competition=competition,
        starters=[LineupPlayer(player_id=pid, nome=names.get(pid, str(pid))) for pid in planned.starts],
        bench=[LineupPlayer(player_id=pid, nome=names.get(pid, str(pid))) for pid in planned.bench],
    )


@router.get("/lineup/plan", response_model=LineupPlan, tags=["lineup"])
def lineup_plan(league_id: int) -> LineupPlan:
    from fantabot.adapters.persistence import database_manager
    from fantabot.adapters.tokens.store import TokenStore
    from fantabot.application.lineup_submit import build_plans
    from fantabot.config import settings
    from fantabot.domain.tokens.crypto import TokenCipher
    from fantabot.domain.tokens.errors import (
        ApiTimeout,
        ApiUnavailable,
        AppKeyRejected,
        TokenError,
        TokenRejected,
    )
    from sqlalchemy.exc import SQLAlchemyError

    from fantabot_app.api.outcomes import because

    key = settings.fantabot_encryption_key
    if not key:
        return LineupPlan(
            found=False,
            outcome="no_credential",
            reason="No encryption key set — connect an account first.",
        )

    # The order is the order the questions arise, and it is load-bearing: a lega with no
    # stored token must not be reported as a network failure, and a network failure must
    # not be reported as a missing lineup.
    try:
        cipher = TokenCipher(key)
    except TokenError as exc:
        # `TokenCipher` raises `KeyMissing`/`KeyMalformed`, both `TokenError`. Named, so a
        # different failure in the constructor is a 500 and not a misleading "not connected".
        return LineupPlan(found=False, outcome="no_credential", reason=str(exc))

    try:
        with database_manager.get_session() as session:
            # The lifted builder, not a third copy of it. This route hand-wrote the same
            # seven reads — `my_team`, `competitions`, `teamLineup_read`, `lineup_settings`,
            # `roster_settings`, `inputs_from_lineup`, `plan_lineups` — which is how it came
            # to read the format from a different place than the command did.
            plans, names, comp = build_plans(TokenStore(session, cipher), league_id, 0)
    except (TokenRejected, AppKeyRejected) as exc:
        # The platform answered, and said no. A different fact from being unable to ask —
        # a rejected token is not going to resolve by reloading the page.
        return LineupPlan(found=False, outcome="refused", reason=str(exc))
    except (ApiTimeout, ApiUnavailable) as exc:
        return LineupPlan(found=False, outcome="unreachable", reason=str(exc))
    except TokenError as exc:
        # Nothing stored, stored under another key, expired, or for another lega. Each
        # says so in its own words and each names something the operator does by hand.
        return LineupPlan(found=False, outcome="no_credential", reason=str(exc))
    except (SQLAlchemyError, OSError) as exc:
        return LineupPlan(found=False, outcome="unreachable", reason=because(exc))

    if not plans:
        # There is a roster and the platform is reachable; no eleven of it is fieldable.
        # `plan_lineups` builds only on natural roles, so this means the rosa genuinely
        # cannot fill a module — not that the request failed.
        return LineupPlan(
            found=False,
            outcome="no_lineup",
            reason="No fieldable lineup for this lega yet — the rosa fills no allowed module.",
        )
    return build_lineup_plan(plans[0], names, comp)


#: Why a submit did not happen, as a screen. Same discipline as `api/outcomes.py`: a route
#: that returns one of a pinned tuple, compared for equality by a test.
#:
#: `not_armed` is a **success** as far as the request is concerned — a dry run is what the
#: caller asked for unless it said otherwise — so it carries the plan it would have sent.
SUBMIT_OUTCOMES = (
    "submitted",
    "not_armed",
    "no_matchday",
    "all_modules_refused",
    "no_credential",
    "refused",
    "unreachable",
)


class SubmitRequest(BaseModel):
    """**`arm` has no default on purpose.** A request that omits it is a 422, not a dry run.

    A default either way is a decision the last request makes for the next one, and the
    property being bought is that the operator who armed it is the one watching. It is never
    stored, never remembered across a reload, and never read from anywhere but this body.
    """

    league_id: int
    arm: bool
    competition: int = 0


class SubmitResult(BaseModel):
    outcome: str
    #: Empty on `submitted`. Names **every** shut lock on `not_armed`, not just the first.
    reason: str = ""
    #: The module that was sent, or — on a dry run — the one that would have been.
    module: str = ""
    matchday: int | None = None
    starters: list[LineupPlayer] = []
    bench: list[LineupPlayer] = []
    #: `True` only when a lineup actually reached the platform.
    submitted: bool = False
    #: The lineup **read back** afterwards: what the platform kept, not what we sent.
    saved_starters: int | None = None
    saved_at: str | None = None
    #: `(module, code)` for each module the platform refused before one stuck — the record
    #: of a `LUP009` walk-down, and empty on a first-try success.
    rejected: list[str] = []
    #: The `mstr` that looks past kickoff. A **warning** carried alongside a submit, never
    #: instead of one: `mstr` is not confirmed to be the lineup deadline.
    past_deadline: str | None = None
    #: Why the confirming read-back failed, when it did. Non-empty means the lineup reached
    #: the platform — `submitted` is `True` — and could not then be read back to prove it.
    #: The outcome deliberately stays `"submitted"`: a reader that does not know this field
    #: says "submitted", which is the true half, where a new outcome string would drop it
    #: into an unknown branch and lose that. An unknown is not a negative.
    unconfirmed: str = ""


#: What the app says when a lock is shut. The *fact* is shared with the CLI
#: (`application.arming`); the wording is local, because there is no `--arm` flag in an HTTP
#: request and a message naming one sends the reader to a terminal they are not using.
APP_SENTENCES = {
    ARM: "the request did not ask to arm",
    AUTO_ACT: "FANTABOT_AUTO_ACT is false",
}


@router.post("/lineup/submit", response_model=SubmitResult, tags=["lineup"])
def lineup_submit(request: SubmitRequest) -> SubmitResult:
    """Build the XI and submit it — **behind two locks, a dry run by default.**

    The eight decisions are `application/lineup_submit.submit_lineup`'s, the same ones
    `fantabot lineup submit` walks. This route chooses a screen for each.
    """
    from fantabot.adapters.persistence import database_manager
    from fantabot.adapters.tokens.store import TokenStore
    from fantabot.application.lineup_submit import (
        ALL_MODULES_REFUSED,
        NO_MATCHDAY,
        NOT_ARMED,
        submit_lineup,
    )
    from fantabot.config import settings
    from fantabot.domain.lineup.errors import LineupError
    from fantabot.domain.tokens.crypto import TokenCipher
    from fantabot.domain.tokens.errors import (
        ApiTimeout,
        ApiUnavailable,
        AppKeyRejected,
        TokenError,
        TokenRejected,
    )
    from sqlalchemy.exc import SQLAlchemyError

    from fantabot_app.api.outcomes import because

    key = settings.fantabot_encryption_key
    if not key:
        return SubmitResult(
            outcome="no_credential",
            reason="No encryption key set — connect an account first.",
        )

    try:
        cipher = TokenCipher(key)
        with database_manager.get_session() as session:
            outcome = submit_lineup(
                TokenStore(session, cipher),
                league_id=request.league_id,
                competition=request.competition,
                arm=request.arm,
                now=_now,
            )
    except (TokenRejected, AppKeyRejected, LineupError) as exc:
        return SubmitResult(outcome="refused", reason=str(exc))
    except (ApiTimeout, ApiUnavailable) as exc:
        return SubmitResult(outcome="unreachable", reason=str(exc))
    except TokenError as exc:
        return SubmitResult(outcome="no_credential", reason=str(exc))
    except (SQLAlchemyError, OSError) as exc:
        return SubmitResult(outcome="unreachable", reason=because(exc))

    plan = outcome.plan
    body = SubmitResult(
        outcome="submitted",
        module=plan.module if plan else "",
        matchday=plan.mday if plan else None,
        starters=[
            LineupPlayer(player_id=pid, nome=outcome.names.get(pid, str(pid)))
            for pid in (plan.starts if plan else [])
        ],
        bench=[
            LineupPlayer(player_id=pid, nome=outcome.names.get(pid, str(pid)))
            for pid in (plan.bench if plan else [])
        ],
        rejected=[f"{module} ({code})" for module, code in outcome.rejected],
        past_deadline=outcome.past_deadline,
    )

    if outcome.refused == NO_MATCHDAY:
        return body.model_copy(
            update={
                "outcome": "no_matchday",
                "reason": (
                    "no matchday context for this competition — the lineup has no saved "
                    "coordinates yet. Try once the matchday opens."
                ),
            }
        )
    if outcome.refused == NOT_ARMED:
        # A dry run is not a failure: it is what was asked for. The plan travels with it.
        return body.model_copy(
            update={"outcome": "not_armed", "reason": outcome.arming.because(APP_SENTENCES)}
        )
    if outcome.refused == ALL_MODULES_REFUSED:
        return body.model_copy(
            update={
                "outcome": "all_modules_refused",
                "reason": "every fieldable module was refused by the platform.",
            }
        )

    return body.model_copy(
        update={
            "submitted": True,
            "unconfirmed": outcome.unconfirmed,
            "saved_starters": len(outcome.saved.get("starts", [])),
            "saved_at": str(outcome.saved.get("ldate")) if outcome.saved.get("ldate") else None,
        }
    )


# -- the scheduled history ----------------------------------------------------------------

#: How many runs one response carries by default, and at most.
DEFAULT_RUNS_LIMIT = 50
MAX_RUNS_LIMIT = 500

#: Older than this and the history says the scheduled job may not be running. The job fires
#: through the day and not overnight, so a normal night leaves roughly nine hours between runs;
#: twelve means a missed day, not a quiet night.
STALE_AFTER_HOURS = 12.0


class LineupRejectionRow(BaseModel):
    """One plan the walk did not keep — `lineup_runs.LineupRejection`, field for field."""

    module: str
    #: `LUP0xx`, or `GUARD <slot> <role> <cell>` for a plan our own guard skipped.
    code: str
    message: str = ""
    #: The XI that was refused, in the platform's slot order.
    starter_ids: list[int] = []
    starters: list[str] = []


class LineupShadowRow(BaseModel):
    """The other model's plan, logged beside the one sent — `lineup_runs.LineupShadow`."""

    model: str
    module: str
    starter_ids: list[int]
    bench_ids: list[int]
    starters: list[str]
    bench: list[str]
    #: E[league points], `3·P(W) + P(D)`: the objective.
    e_pts: float
    #: P(win), P(draw), P(loss).
    p_wdl: tuple[float, float, float]
    e_fp: float
    sd: float
    cuts: list[str] = []


class LineupRunRow(BaseModel):
    """One scheduled run — `adapters/files/lineup_runs.LineupRun`, field for field.

    Built with `LineupRunRow(**asdict(run))` rather than a hand-written mapping: that mapping
    is the seam the room journal's three dropped keys came through. That alone does not
    close it — pydantic *ignores* a key the model does not declare, so a field the writer
    gains is dropped here without a word. `test_lineup_runs_route` pins the field sets equal,
    for this row and both nested ones.
    """

    at: str
    league: int
    scheduled: bool
    #: `submitted`, `unconfirmed`, `skipped` or `failed` — decided once, by
    #: `application/lineup_submit.run_record`. This screen colours it and never re-derives it.
    status: str
    code: str = ""
    detail: str = ""
    module: str = ""
    matchday: int | None = None
    serie_a_matchday: int | None = None
    starters: list[str] = []
    bench: list[str] = []
    rejected: list[str] = []
    # --- record v2: each defaults to what a v1 line did not record ---
    #: The ids are what grading reads; the names above are for the eye.
    starter_ids: list[int] = []
    bench_ids: list[int] = []
    competition: int | None = None
    tid: int | None = None
    #: The model that chose what was sent. `""` on a v1 line, which never said.
    model: str = ""
    #: Why the named model was not the one used, when it was not.
    fallback: str = ""
    warnings: list[str] = []
    rejections: list[LineupRejectionRow] = []
    shadow: LineupShadowRow | None = None


class LineupRuns(BaseModel):
    ok: bool
    #: The file actually read, so a screen that shows nothing says where it looked.
    path: str
    exists: bool
    total: int = 0
    skipped: int = 0
    runs: list[LineupRunRow] = []
    error: str | None = None
    last_at: str | None = None
    last_age_hours: float | None = None
    #: The newest run is older than `STALE_AFTER_HOURS`. A job that stopped running writes no
    #: row at all, so without this a list of green rows from last week reads as fine.
    stale: bool = False


def read_lineup_runs(
    path: Path, *, now: datetime, limit: int = DEFAULT_RUNS_LIMIT
) -> LineupRuns:
    """The scheduled lineup's history at `path`, newest first, and how old the newest run is.

    Three answers a screen must tell apart, as the room journal's reader does: **missing** is
    "no scheduled run recorded yet"; **unreadable** is its own error, because rendering a
    directory-where-a-file-should-be as "nothing yet" sends the operator looking for a path
    that is already right; and **present**, with the age of its newest row. `now` is this
    endpoint's `_now()`, so the app's lineup surface still reads the clock in one place.
    """
    import dataclasses

    from fantabot.adapters.files.lineup_runs import read_runs

    limit = max(1, min(limit, MAX_RUNS_LIMIT))
    shown = str(path)
    existed = path.exists()
    try:
        runs, skipped = read_runs(path)
    except OSError as exc:  # a directory, a permission, a vanished volume
        return LineupRuns(ok=False, path=shown, exists=True, error=type(exc).__name__)
    if not existed:
        return LineupRuns(ok=True, path=shown, exists=False)

    newest = runs[0] if runs else None
    age: float | None = None
    if newest is not None:
        try:
            age = (now.astimezone() - datetime.fromisoformat(newest.at)).total_seconds() / 3600
        except (ValueError, TypeError):
            age = None
    return LineupRuns(
        ok=True,
        path=shown,
        exists=True,
        total=len(runs),
        skipped=skipped,
        runs=[LineupRunRow(**dataclasses.asdict(run)) for run in runs[:limit]],
        last_at=newest.at if newest else None,
        last_age_hours=None if age is None else round(age, 1),
        stale=age is not None and age > STALE_AFTER_HOURS,
    )


@router.get("/lineup/runs", response_model=LineupRuns, tags=["lineup"])
def lineup_runs(limit: int = DEFAULT_RUNS_LIMIT) -> LineupRuns:
    """What the scheduled lineup job did, read back. **The app writes nothing here** and
    offers no control over the job: turning it on and off lives in `.env` and the `launchd`
    plist, by the operator's choice, and `arming.py` exists so no arming decision is stored.
    """
    from fantabot.config import lineup_runs_path

    return read_lineup_runs(lineup_runs_path(), now=_now(), limit=limit)


# -- what the platform has saved right now ------------------------------------------------


class CurrentLineup(BaseModel):
    """`fantabot lineup show`, as a value.

    **Ids, not names.** The command prints ids and so does this: naming them would mean
    running `build_plans` — seven live reads and a solve — to annotate a read that has
    already answered, and it would fail for reasons that have nothing to do with the saved
    lineup. The page holds the plan for the same lega and joins the names client-side,
    falling back to the id, which costs nothing it was not already paying.
    """

    #: One of `api/outcomes.LINEUP_CURRENT_OUTCOMES`.
    outcome: str
    reason: str = ""
    module: str = ""
    starters: list[int] = []
    bench: list[int] = []


@router.get("/lineup/current", response_model=CurrentLineup, tags=["lineup"])
def lineup_current(league_id: int, competition: int) -> CurrentLineup:
    """The lineup the platform currently holds for one competition. **Read-only.**

    A different question from `GET /lineup/plan`, and the difference is the point: that one
    asks what we *should* field, this asks what is *saved*. On a matchday where a submit was
    refused — an unfieldable module, a deadline already past — the two answers differ, and
    that is exactly when an operator needs to see both.

    `competition` is required rather than resolved. A lineup belongs to a competition, and
    picking one would answer a question nobody asked; `lineup show` refuses without one for
    the same reason.
    """
    from fantabot.domain.tokens.errors import (
        ApiTimeout,
        ApiUnavailable,
        AppKeyRejected,
        TokenError,
        TokenRejected,
    )
    from sqlalchemy.exc import SQLAlchemyError

    from fantabot_app.api.outcomes import because

    try:
        with _open_store() as store:
            body = teamLineup_read(league_id, competition, store=store)
    except (TokenRejected, AppKeyRejected) as exc:
        # Both, and before the `TokenError` clause, because `AppKeyRejected` **is** a
        # `TokenError`: caught only as one, the same platform rejection reads as
        # `no_credential` here and `refused` on `GET /lineup/plan` — two routes, one
        # credential, two different remedies on screen, which is the thing `outcomes.py`
        # exists to prevent.
        return CurrentLineup(outcome="refused", reason=str(exc))
    except (ApiTimeout, ApiUnavailable) as exc:
        return CurrentLineup(outcome="unreachable", reason=str(exc))
    except TokenError as exc:
        return CurrentLineup(outcome="no_credential", reason=str(exc))
    except (SQLAlchemyError, OSError) as exc:
        return CurrentLineup(outcome="unreachable", reason=because(exc))

    dto = body.get("teamLineupDto") or {}
    if not dto:
        return CurrentLineup(
            outcome="no_lineup",
            reason="no lineup has been saved for this competition yet.",
        )
    return CurrentLineup(
        outcome="read",
        module=str(dto.get("mdl", "")),
        starters=[int(p) for p in dto.get("starts", [])],
        bench=[int(p) for p in dto.get("bench", [])],
    )

File: app/frontend/src/app/pages/lineup/lineup.ts
import {
  ChangeDetectionStrategy,
  Component,
  DestroyRef,
  ElementRef,
  Injector,
  OnInit,
  afterNextRender,
  computed,
  inject,
  signal,
  viewChild,
} from '@angular/core';
import { takeUntilDestroyed } from '@angular/core/rxjs-interop';
import { MatButtonModule } from '@angular/material/button';
import { MatButtonToggleModule } from '@angular/material/button-toggle';
import { MatCardModule } from '@angular/material/card';
import { MatListModule } from '@angular/material/list';
import { MatProgressBarModule } from '@angular/material/progress-bar';
import { LucideAngularModule } from 'lucide-angular';

import { LegaService } from '../../core/api/lega.service';
import { LineupService } from '../../core/api/lineup.service';
import { LegaOverview } from '../../core/models/lega';
import { CurrentLineup, LineupPlan, LineupRuns, SubmitResult } from '../../core/models/lineup';

@Component({
  selector: 'app-lineup',
  imports: [
    LucideAngularModule,
    MatButtonModule,
    MatButtonToggleModule,
    MatCardModule,
    MatListModule,
    MatProgressBarModule,
  ],
  templateUrl: './lineup.html',
  styleUrl: './lineup.scss',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class LineupComponent implements OnInit {
  private readonly lega = inject(LegaService);
  private readonly lineup = inject(LineupService);
  private readonly destroyRef = inject(DestroyRef);
  private readonly injector = inject(Injector);

  /** Where focus goes when a submit lands — see `focusResult`. */
  private readonly resultBlock = viewChild<ElementRef<HTMLElement>>('resultBlock');

  readonly leagues = signal<LegaOverview[]>([]);
  readonly selectedId = signal<number | null>(null);
  readonly plan = signal<LineupPlan | null>(null);
  readonly loading = signal(true);
  readonly planLoading = signal(false);
  readonly errorMsg = signal<string | null>(null);

  /**
   * The dry run's result, and the gate on arming.
   *
   * **Arming requires a second, explicit act on a dry run the operator has just seen.** Not
   * a checkbox — a checkbox pre-ticked from last time is not a second act, and a page can
   * be restored by a session manager with its form state intact. Selecting another lega
   * clears it, so the run on screen is always the run that would be armed.
   */
  readonly dryRun = signal<SubmitResult | null>(null);
  readonly submitting = signal(false);
  readonly result = signal<SubmitResult | null>(null);

  /**
   * Whether a dry run for the *currently selected* lega is on screen — and so whether the
   * second act is offered at all.
   *
   * Deliberately says nothing about `submitting`. The arm button used to live inside
   * `canArm`, which meant clicking it destroyed it: `arm()` sets `submitting` as its first
   * act, the block stopped matching, and the focused element was removed from under the
   * keyboard on the one irreversible action in the app. The button now stays mounted and
   * goes disabled instead.
   */
  readonly armVisible = computed(() => this.dryRun()?.outcome === 'not_armed');

  /** True only while a dry run is on screen and nothing is in flight. */
  readonly canArm = computed(() => this.armVisible() && !this.submitting());

  /**
   * The scheduled job's history — what `launchd` did, read back. Read-only on purpose: the
   * operator turns the job on and off in `.env` and the plist, never here, so there is no
   * control on this page to go with it.
   */
  readonly runs = signal<LineupRuns | null>(null);
  readonly runsError = signal<string | null>(null);

  /**
   * What the platform has saved right now — `fantabot lineup show`, on a screen.
   *
   * A different question from the plan beside it, and the difference is the point: the plan
   * is what we *should* field, this is what is *saved*. On a matchday where a submit was
   * refused — an unfieldable module, a deadline already past — the two answers differ, and
   * the operator is looking at a page that would otherwise show only one of them.
   *
   * Fetched with the plan rather than behind a button: it is one small read, and a saved
   * lineup that disagrees with the plan is exactly the thing nobody would think to ask for.
   */
  readonly current = signal<CurrentLineup | null>(null);
  readonly currentError = signal<string | null>(null);

  /**
   * The saved lineup with names joined from the plan this page already holds.
   *
   * The server returns ids: naming them there would mean running the whole plan to annotate
   * a read that has already answered, and failing for reasons that have nothing to do with
   * the save. Here the join is free — and it falls back to the id, which is what a player
   * outside the planned roster looks like and is worth seeing as such.
   */
  readonly currentNamed = computed(() => {
    const saved = this.current();
    if (!saved || saved.outcome !== 'read') return null;
    const names = new Map(
      [...(this.plan()?.starters ?? []), ...(this.plan()?.bench ?? [])].map((p) => [
        p.player_id,
        p.nome,
      ]),
    );
    const name = (id: number) => names.get(id) ?? String(id);
    return {
      module: saved.module,
      starters: saved.starters.map(name),
      bench: saved.bench.map(name),
    };
  });

  ngOnInit(): void {
    this.loadLeagues();
    this.loadRuns();
  }

  loadRuns(): void {
    this.runsError.set(null);
    this.lineup
      .getRuns()
      .pipe(takeUntilDestroyed(this.destroyRef))
      .subscribe({
        next: (runs) => this.runs.set(runs),
        error: () => this.runsError.set('Could not reach the API for the scheduled history.'),
      });
  }

  loadLeagues(): void {
    this.loading.set(true);
    this.errorMsg.set(null);
    this.lega
      .getLeagues()
      .pipe(takeUntilDestroyed(this.destroyRef))
      .subscribe({
        next: (leagues) => {
          this.leagues.set(leagues);
          this.loading.set(false);
          if (leagues.length) this.select(leagues[0].league_id);
        },
        error: () => {
          this.errorMsg.set('Could not reach the API.');
          this.loading.set(false);
        },
      });
  }

  select(leagueId: number): void {
    this.selectedId.set(leagueId);
    this.plan.set(null);
    // A dry run belongs to one lega. Carrying it across would let an operator arm a lineup
    // they never saw — the whole property this gate exists for.
    this.dryRun.set(null);
    this.result.set(null);
    this.current.set(null);
    this.currentError.set(null);
    this.fetchPlan(leagueId);
  }

  /**
   * Re-read the plan without clearing what is on screen.
   *
   * Split from `select` because arming reloads the plan and **must not** wipe the result
   * the operator has just been handed — the read-back is the evidence that the platform
   * kept it, and it was being erased a frame after it arrived.
   */
  private fetchPlan(leagueId: number, { quiet = false } = {}): void {
    // `quiet` skips the loading state. The template swaps the whole panel for a skeleton
    // while `planLoading` is set, so a background refresh after arming would blank the
    // read-back a frame after the operator was handed it — the one thing on that screen
    // that is evidence.
    if (!quiet) this.planLoading.set(true);
    this.lineup
      .getPlan(leagueId)
      .pipe(takeUntilDestroyed(this.destroyRef))
      .subscribe({
        next: (plan) => {
          this.plan.set(plan);
          this.planLoading.set(false);
          // Only once the plan has named a competition. Asking for the saved lineup of a
          // competition nobody resolved would be a second guess about which one, and the
          // two screens would stop being comparable — which is the whole reason for both.
          //
          // `typeof === 'number'`, not `!== null`: a server that predates this field answers
          // `undefined`, which is not null, and the page would fire a request for
          // `competition=undefined`. That is the second-request hazard the standing list
          // already records — an outstanding request fails `httpMock.verify()` and a failing
          // `afterEach` leaves the TestBed up, which once took down three unrelated specs.
          if (typeof plan.competition === 'number') {
            this.fetchCurrent(leagueId, plan.competition);
          }
        },
        error: () => this.planLoading.set(false),
      });
  }

  /**
   * Read what the platform has saved for the competition the plan named.
   *
   * A failed read is reported and nothing else: the plan beside it is still the answer to
   * its own question, and blanking the page over the secondary read would lose it.
   */
  private fetchCurrent(leagueId: number, competition: number): void {
    this.currentError.set(null);
    this.lineup
      .getCurrent(leagueId, competition)
      .pipe(takeUntilDestroyed(this.destroyRef))
      .subscribe({
        next: (saved) => {
          this.current.set(saved);
          // The server's own sentence, never one composed here — five outcomes, and only
          // one of them is fixed by reloading.
          this.currentError.set(saved.outcome === 'read' ? null : saved.reason);
        },
        error: () => this.currentError.set('Could not reach the API for the saved lineup.'),
      });
  }

  /**
   * Ask what would be sent. Never arms — `arm: false` is passed explicitly, every time.
   */
  runDry(): void {
    const leagueId = this.selectedId();
    if (leagueId === null || this.submitting()) return;
    this.submitting.set(true);
    this.result.set(null);
    this.lineup
      .submit(leagueId, false)
      .pipe(takeUntilDestroyed(this.destroyRef))
      .subscribe({
        next: (outcome) => {
          this.dryRun.set(outcome);
          this.submitting.set(false);
        },
        error: () => {
          this.errorMsg.set('Could not reach the API.');
          this.submitting.set(false);
        },
      });
  }

  /**
   * Submit for real — the second act.
   *
   * Refuses unless a dry run for this lega is on screen. That is enforced here and not only
   * by hiding the button: a disabled control is a suggestion, and this is the one call in
   * the app that spends a matchday.
   */
  arm(): void {
    const leagueId = this.selectedId();
    if (leagueId === null || !this.canArm()) return;
    this.submitting.set(true);
    this.lineup
      .submit(leagueId, true)
      .pipe(takeUntilDestroyed(this.destroyRef))
      .subscribe({
        next: (outcome) => {
          this.result.set(outcome);
          // Spent. The next arm needs its own dry run, on whatever the roster is now.
          this.dryRun.set(null);
          this.submitting.set(false);
          this.fetchPlan(leagueId, { quiet: true });
          this.focusResult();
        },
        error: () => {
          this.errorMsg.set('Could not reach the API.');
          this.submitting.set(false);
        },
      });
  }

  /**
   * Land focus on the outcome the submit produced.
   *
   * The arm button is disabled while the call is in flight and the whole block goes away
   * once the dry run is spent, so by the time the result is on screen there is nothing
   * left for focus to sit on — the browser drops it on `<body>` and the keyboard restarts
   * at the top of the page, on the one action that cannot be taken back.
   * `afterNextRender`, not a direct call: `result.set` is what *creates* the element being
   * focused, so it does not exist until the view has been refreshed. Same idea as
   * `accounts.ts::focusSection`, which can focus directly only because its heading was
   * already in the DOM.
   */
  private focusResult(): void {
    afterNextRender(() => this.resultBlock()?.nativeElement.focus(), { injector: this.injector });
  }
}
